app/assets/javascripts/discourse/app/routes/new-message.js

```diff
@@ -12,49 +12,69 @@ export default DiscourseRoute.extend({
-
+    const userName = params.username;
     const groupName = params.groupname || params.group_name;
 
     if (this.currentUser) {
-      this.router
-        .replaceWith("discovery.latest")
-        .followRedirects()
-        .then(() => {
-          if (params.username) {
-            this.composer.openNewMessage({
-              recipients: params.username,
-              title: params.title,
-              body: params.body,
-            });
-          } else if (groupName) {
-            // send a message to a group
-            Group.messageable(groupName)
-              .then((result) => {
-                if (result.messageable) {
-                  next(() =>
-                    this.composer.openNewMessage({
-                      recipients: groupName,
-                      title: params.title,
-                      body: params.body,
-                    })
-                  );
-                } else {
-                  this.dialog.alert(
-                    I18n.t("composer.cant_send_pm", { username: groupName })
-                  );
-                }
-              })
-              .catch(() => this.dialog.alert(I18n.t("generic_error")));
-          } else {
-            this.composer.openNewMessage({
-              title: params.title,
-              body: params.body,
-            });
-          }
-        });
+      if (transition.from) {
+        transition.abort();
+
+        if (userName) {
+          this.openComposer(transition, userName);
+        } else if (groupName) {
+          // send a message to a group
+          Group.messageable(groupName)
+            .then((result) => {
+              if (result.messageable) {
+                this.openComposer(transition, groupName);
+              } else {
+                this.dialog.alert(
+                  I18n.t("composer.cant_send_pm", { username: groupName })
+                );
+              }
+            })
+            .catch(() => this.dialog.alert(I18n.t("generic_error")));
+        } else {
+          this.openComposer(transition);
+        }
+      } else {
+        this.router
+          .replaceWith("discovery.latest")
+          .followRedirects()
+          .then(() => {
+            if (userName) {
+              this.openComposer(transition, userName);
+            } else if (groupName) {
+              // send a message to a group
+              Group.messageable(groupName)
+                .then((result) => {
+                  if (result.messageable) {
+                    this.openComposer(transition, groupName);
+                  } else {
+                    this.dialog.alert(
+                      I18n.t("composer.cant_send_pm", { username: groupName })
+                    );
+                  }
+                })
+                .catch(() => this.dialog.alert(I18n.t("generic_error")));
+            } else {
+              this.openComposer(transition);
+            }
+          });
+      }
```

Comment:
It is hard to follow this component, and I felt overwhelmed when I first looked at it.

Moving it to an early return pattern instead of those else if branches would not make it easier to understand and maintain? 

this is this same piece _without_ the changes but with this refactor
```js

    if (!this.currentUser) {
      cookie("destination_url", window.location.href);
      this.router.replaceWith("login");
      return;
    }

    this.router
      .replaceWith("discovery.latest")
      .followRedirects()
      .then(() => {
        if (params.username) {
          return this.composer.openNewMessage({
            recipients: params.username,
            title: params.title,
            body: params.body,
          });
        }

        if (groupName) {
          return Group.messageable(groupName).then((result) => {
            if (!result.messageable) {
              return this.dialog.alert(
                I18n.t("composer.cant_send_pm", { username: groupName })
              );
            }
            return next(() =>
              this.composer.openNewMessage({
                recipients: groupName,
                title: params.title,
                body: params.body,
              })
            );
          });
        }
        return this.composer.openNewMessage({
          title: params.title,
          body: params.body,
        });
      });

```

Reply:
Good idea @Grubba27, updated the PR to prefer early return pattern.